File: src/Core/IoBoundary.fs
namespace Zeta.Core

open System.Runtime.CompilerServices

/// Typed inside/outside boundary for the signed-ledger-to-grow-only transition.
///
/// `Inside` may hold a signed Z-set ledger: emits and retracts can cancel before
/// observation. `Outside` holds only the fused G-set view: monotone presence,
/// with multiplicity, negative evidence, and the path that produced the result
/// kept behind the boundary.
[<RequireQualifiedAccess>]
module IoBoundary =

    [<Struct; IsReadOnly>]
    type Inside<'K when 'K : comparison> =
        val internal Ledger: ZSet<'K>
        internal new(ledger: ZSet<'K>) = { Ledger = ledger }

    [<Struct; IsReadOnly>]
    type Outside<'K when 'K : comparison> =
        val internal View: GSet<'K>
        internal new(view: GSet<'K>) = { View = view }

    let emptyInside<'K when 'K : comparison> : Inside<'K> =
        Inside ZSet.empty

    let emptyOutside<'K when 'K : comparison> : Outside<'K> =
        Outside GSet.empty

    /// Enter the boundary with an already-composed signed ledger.
    let input (z: ZSet<'K>) : Inside<'K> =
        Inside z

    /// Enter the boundary with add-only genesis facts.
    let genesis (keys: 'K seq) : Inside<'K> =
        keys |> ZSet.ofKeys |> Inside

    /// One positive internal event.
    let emit (key: 'K) : Inside<'K> =
        ZSet.singleton key 1L |> Inside

    /// One negative internal event.
    let retract (key: 'K) : Inside<'K> =
        ZSet.singleton key -1L |> Inside

    /// Compose signed interiors before any exterior observation occurs.
    let compose (left: Inside<'K>) (right: Inside<'K>) : Inside<'K> =
        ZSet.add left.Ledger right.Ledger |> Inside

    let composeAll (insides: Inside<'K> seq) : Inside<'K> =
        let mutable acc = ZSet.empty
        for inside in insides do
            acc <- ZSet.add acc inside.Ledger
        Inside acc

    /// Cross the I/O boundary: only positive support becomes exterior identity.
    let fuse (inside: Inside<'K>) : Outside<'K> =
        inside.Ledger |> FusionReconstruction.fuse |> Outside

    /// Leave the boundary with the public grow-only view.
    let output (outside: Outside<'K>) : GSet<'K> =
        outside.View

    let toArray (outside: Outside<'K>) : 'K[] =
        GSet.toArray outside.View

    let toList (outside: Outside<'K>) : 'K list =
        GSet.toList outside.View

    let contains (key: 'K) (outside: Outside<'K>) : bool =
        GSet.contains key outside.View

    let count (outside: Outside<'K>) : int =
        GSet.count outside.View

    let isEmpty (outside: Outside<'K>) : bool =
        GSet.isEmpty outside.View

File: src/Core.TypeScript/io-boundary/io-boundary.ts
/**
 * io-boundary.ts — typed inside/outside boundary for the signed-ledger-to-grow-only transition.
 *
 * `Inside` holds a signed Z-set ledger where emissions and retractions can cancel before
 * observation. `Outside` holds only the fused G-set view: monotone presence, with
 * multiplicity, negative evidence, and the path that produced the result kept behind the boundary.
 *
 * Mirroring the F# canonical shape (src/Core/IoBoundary.fs).
 */

import * as ZSet from "../z-set/z-set";
import * as GSet from "../g-set/g-set";

// Unexported symbols to keep internal states inaccessible from outside this file
export const ledgerSymbol = Symbol("ledger");
export const viewSymbol = Symbol("view");

/**
 * Inside boundary wrapper: holds a signed Z-set ledger.
 */
export class Inside<K> {
  /** @internal */
  readonly [ledgerSymbol]: ZSet.ZSet<K>;

  constructor(ledger: ZSet.ZSet<K>) {
    this[ledgerSymbol] = ledger;
  }
}

/**
 * Outside boundary wrapper: holds the public fused G-set view.
 */
export class Outside<K> {
  /** @internal */
  readonly [viewSymbol]: GSet.GSet<K>;

  constructor(view: GSet.GSet<K>) {
    this[viewSymbol] = view;
  }
}

/** The empty Inside boundary. */
export function emptyInside<K>(): Inside<K> {
  return new Inside(ZSet.empty<K>());
}

/** The empty Outside boundary. */
export function emptyOutside<K>(): Outside<K> {
  return new Outside(GSet.empty<K>());
}

/** Enter the boundary with an already-composed signed ledger. */
export function input<K>(z: ZSet.ZSet<K>): Inside<K> {
  return new Inside(z);
}

/** Enter the boundary with add-only genesis facts. */
export function genesis<K>(compare: ZSet.Compare<K>, keys: readonly K[]): Inside<K> {
  return new Inside(ZSet.ofArray(compare, keys));
}

/** One positive internal event. */
export function emit<K>(key: K): Inside<K> {
  return new Inside(ZSet.singleton(key, 1));
}

/** One negative internal event. */
export function retract<K>(key: K): Inside<K> {
  return new Inside(ZSet.singleton(key, -1));
}

/** Compose signed interiors before any exterior observation occurs. */
export function compose<K>(compare: ZSet.Compare<K>, left: Inside<K>, right: Inside<K>): Inside<K> {
  return new Inside(ZSet.union(compare, left[ledgerSymbol], right[ledgerSymbol]));
}

/** Compose a sequence of signed interiors. */
export function composeAll<K>(compare: ZSet.Compare<K>, insides: readonly Inside<K>[]): Inside<K> {
  let acc = ZSet.empty<K>();
  for (const inside of insides) {
    acc = ZSet.union(compare, acc, inside[ledgerSymbol]);
  }
  return new Inside(acc);
}

/** Cross the I/O boundary: only positive support becomes exterior identity. */
export function fuse<K>(inside: Inside<K>): Outside<K> {
  const ledger = inside[ledgerSymbol];
  const view: K[] = [];
  for (const entry of ledger) {
    if (entry.w > 0) {
      view.push(entry.e);
    }
  }
  // The resulting view is guaranteed to be GSet-compliant (sorted and unique)
  // because the source ZSet entries are already sorted and unique by key.
  return new Outside(view);
}

/** Leave the boundary with the public grow-only view. */
export function output<K>(outside: Outside<K>): GSet.GSet<K> {
  return outside[viewSymbol];
}

/** Return the exterior view as a fresh array. */
export function toArray<K>(outside: Outside<K>): K[] {
  return [...outside[viewSymbol]];
}

/** Return the exterior view as a fresh list array (matching F# toList). */
export function toList<K>(outside: Outside<K>): K[] {
  return [...outside[viewSymbol]];
}

/** Membership: check if the key is present in the GSet view. */
export function contains<K>(compare: ZSet.Compare<K>, key: K, outside: Outside<K>): boolean {
  return GSet.contains(compare, outside[viewSymbol], key);
}

/** Count of elements in the exterior view. */
export function count<K>(outside: Outside<K>): number {
  return outside[viewSymbol].length;
}

/** Check if the exterior view is empty. */
export function isEmpty<K>(outside: Outside<K>): boolean {
  return outside[viewSymbol].length === 0;
}
